compiler/rustc_hir_analysis/src/collect/type_of.rs

```diff
@@ -619,7 +618,16 @@ fn infer_placeholder_type<'a>(
             }
             diag.emit()
         });
-    Ty::new_error(tcx, guar)
+
+    // Typeck returns regions as erased. We can't deal with erased regions here though, so we
+    // turn them into `&'static`, which is *generally* correct for statics and consts.
+    // Assoc consts can reference generic lifetimes from the parent generics, but treating them
+    // as static is unlikely to cause issues.
+    let ty = tcx.fold_regions(ty, |region, _| match region.kind() {
+        ty::ReErased => tcx.lifetimes.re_static,
```

Comment:
```suggestion
        ty::ReErased => ty::Region::new_error(tcx, guar),
```

would also fix this issue without requiring any of the other changes, because then the error type is tainted, without replacing any of the types. So you'd still get the suggestions, but const eval won't attempt to run.